### PR TITLE
add testId attribute

### LINKED ISSUE #262: Separate test id into separate widget attribute
**Is your feature request related to a problem? Please describe.**
Registering bindings for widgets with IDs is expensive because all atoms must be re-evaluated.

**Describe the solution you'd like**
Separate testId into a separate attribute

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by anserwaseem**
## Issue ticket number and link

Closes #262

## Checklist before requesting a review

- [ ] I have performed a self-review of my code
- [ ] I have added tests

### CODE REVIEW

File: packages/framework/src/hooks/useWidgetId.ts [hunk 1]
```diff
@@ -8,10 +8,11 @@ import { useCustomScope } from "./useCustomScope";
 
 export const useWidgetId = (
   id?: Expression<string>,
+  testId?: Expression<string>,
 ): { resolvedWidgetId: string; rootRef: RefCallback<never> } => {
   const customScope = useCustomScope();
   const resolvedWidgetId = useMemo<string>(() => {
-    let workingId = id;
+    let workingId = id || testId;
```
Comment (evshi) [inline packages/framework/src/hooks/useWidgetId.ts:15]: testId is not a substitute for id, it should only be used for the html attr

File: packages/framework/src/hooks/useWidgetId.ts [hunk 2]
```diff
@@ -26,12 +27,12 @@ export const useWidgetId = (
       );
     }
     return generateRandomString(6);
-  }, [customScope, id]);
+  }, [customScope, id, testId]);
 
   const rootRef = useCallback(
     (node: never) => {
       if (node && "setAttribute" in node) {
-        (node as HTMLElement).setAttribute("data-testid", resolvedWidgetId);
```
Comment (evshi) [inline packages/framework/src/hooks/useWidgetId.ts:34]: keep this the same

File: packages/framework/src/hooks/useRegisterBindings.ts [hunk 3]
```diff
@@ -20,8 +20,9 @@ export const useRegisterBindings = <T extends Record<string, unknown>>(
   values: T,
   id?: string,
   methods?: InvokableMethods,
+  testId?: string,
```
Comment (evshi) [inline packages/framework/src/hooks/useRegisterBindings.ts:23]: This doesn't need a separate arg, just pull it from the values

File: packages/framework/src/hooks/useWidgetId.ts (hunk 2 above)
Comment (anserwaseem) [reply packages/framework/src/hooks/useWidgetId.ts:34]: should we fallback to `resolvedWidgetId` if neither id nor the testId is present?

**Review by evshi — APPROVED** [review]
To confirm, you see the data-testid attribute on the HTML element corresponding to the Row widget?

**Comment by anserwaseem** [comment]
> To confirm, you see the data-testid attribute on the HTML element corresponding to the Row widget?

Yes.

Screenshot 2023-12-28 at 3 36 16 PM

**Pull request merged**